### ISSUE TITLE
FieldMask JSON serialization improvement proposal

### ISSUE THREAD

**Issue by antongrbin** [protocolbuffers/protobuf#12524]
**What language does this apply to?**
This feature request concerns FieldMask JSON serialization.

**Describe the problem you are trying to solve.**
We are using FieldMasks as described in [https://google.aip.dev/161](https://google.aip.dev/161). We understand that the semantics described in AIP are slightly different than the original semantics described in the [field_mask.proto](https://github.com/protocolbuffers/protobuf/blob/983dac24c34f07a278d45fdfd30348775d95d12a/src/google/protobuf/field_mask.proto#L43). Some differences were discussed in https://github.com/aip-dev/google.aip.dev/issues/673 and https://github.com/aip-dev/google.aip.dev/issues/1013.

The differences in semantics are not ideal, but also not a deal-breaker since each implementation can decide on the specifics or give config options.

The wall that we hit is with [https://google.aip.dev/161#map-fields](https://google.aip.dev/161#map-fields) in REST JSON services. The current protobuf FieldMask JSON serialization tries to `camelCase` all fields in the mask and fails if uppercase letter is encountered, for example:
* [FieldMaskUtil.java](https://github.com/protocolbuffers/protobuf/blob/983dac24c34f07a278d45fdfd30348775d95d12a/java/util/src/main/java/com/google/protobuf/util/FieldMaskUtil.java#L176)
* [field_mask.py](https://github.com/protocolbuffers/protobuf/blob/983dac24c34f07a278d45fdfd30348775d95d12a/python/google/protobuf/internal/field_mask.py#L41)

This means that it's not possible to implement [https://google.aip.dev/161#map-fields](https://google.aip.dev/161#map-fields) that extends `FieldMask` semantics to traverse into map keys. For map keys that contain an uppercase letters or special characters, we get an exception during JSON serialization of `google.protobuf.FieldMask`.

**Describe the solution you'd like**
The mentioned AIP suggests a solution:
> Field masks should support string keys that contain characters that are problematic for the field mask syntax, using the backtick character.

We could extend JSON ser/de of FieldMasks to enable backtick quotation. For quoted fields, we wouldn't attempt camelCase/snake_case conversion. I found [one implementation that implements this](https://pkg.go.dev/go.chromium.org/luci/common/proto/mask).

**Describe alternatives you've considered**
Workarounds that we considered are:
* use binary serialization
* since we don't use protobuf field mask utils, we don't need to use `google.protobuf.FieldMask` but a custom similar message

**Additional context**
If you agree with the approach, I can start by sending a PR that adds a recommended conformance test which asks JSON ser/de of FieldMask not to fail on quoted fields with special characters.

**Comment by jskeet** [comment]
Hmm... will give this some thought tomorrow (UK time). My main concern will be compatibility. Definitely happy to see a conformance test to start with, as a good way of documenting the proposed changes.

**Comment by antongrbin** [comment]
@jskeet thank you for your time. As discussed, I created conformance tests: https://github.com/protocolbuffers/protobuf/pull/12551

In the process of doing so I observed that map keys containing a `.` character are a problem for binary serialization as well. A reader can't unambiguously read `"map_field.anton.grbin"`. It feels like FieldMask without backticks is not ready to support map keys.

I found a PR that introduced the AIP: https://github.com/aip-dev/google.aip.dev/pull/701 and you were part of the review, so I am confident that I am in good hands here to get the learnings here :)

**Comment by jskeet** [comment]
Thanks, I'll look carefully through the PR. It's likely that I'll need to write an internal proposal doc for changes and evaluate the impact... basically this could take quite a while, I'm afraid. But hopefully we'll get there.

**Comment by jskeet** [comment]
This bit surprises me though:

> In the process of doing so I observed that map keys containing a . character are a problem for binary serialization as well. A reader can't unambiguously read "map_field.anton.grbin".

Could you clarify exactly what you mean by this? Binary serialization of a FieldMask should just be "normal protobuf binary serialization" - there's no special-casing here; a FieldMask just contains a list of strings.

**Comment by antongrbin** [comment]
Extending the example from here: https://google.aip.dev/161#map-fields

```protobuf
message Book {
  // Metadata about the review.
  message Review {
    string bar = 1;
    string text = 2;
  }

  // A map of string review_id to Review object.
  map<string, Review> reviews = 2;
}
```

Let's imagine two hypothetical requests:
1) Insert a new review entry under key `foo.bar`
2) Update the existing review entry under the key `foo` to change the field `bar` to something

I believe that FieldMasks in both 1 and 2 will be equal. In text proto: `paths: "reviews.foo.bar"`.

The problem is that `google.protobuf.FieldMask` uses `repeated string paths = 1;` where each path is a further concatenation of elements with a delimiter `.`.

With the proposed quoting, the two FieldMask will become distinct. In text proto:
```
// request 1
{ paths: "reviews.`foo.bar`" }

// request 2
{ paths: "reviews.`foo`.bar" }
```

**Comment by jskeet** [comment]
Right - but that doesn't indicate "a problem for binary serialization" to me - it indicates a concern around the *semantics* of field masks.

Fundamentally, if I can serialize the FieldMask and deserialize it again, and get the same value that I started with, then the *serialization* part works. And that's what I'd expect to always be the case for binary serialization, because there's no custom code.

If the concern is "FieldMask semantics for map fields need work" then I'm completely on-board.

**Comment by antongrbin** [comment]
Got it! I agree that there is nothing wrong with binary serialization and FieldMasks!

**Comment by jskeet** [comment]
Starting to write a doc on this now. Please could you clarify this statement:

> The current protobuf FieldMask JSON serialization tries to camelCase all fields in the mask and fails if uppercase letter is encountered, for example

Does serialization throw an exception, or does it produce an inappropriate result?

**Comment by antongrbin** [comment]
I looked into 3 implementations:
* Python, serialization raises an exception if uppercase letter is encountered [code pointer](https://github.com/protocolbuffers/protobuf/blob/983dac24c34f07a278d45fdfd30348775d95d12a/python/google/protobuf/internal/field_mask.py#L148)
* C++, serialization returns false (gives up) if uppercase is encountered [code pointer](https://github.com/protocolbuffers/protobuf/blob/983dac24c34f07a278d45fdfd30348775d95d12a/src/google/protobuf/util/field_mask_util.cc#L77)
* Java, serialization produces inappropriate result - [tested here](https://github.com/protocolbuffers/protobuf/pull/12566/files)

Hope this helps!

**Comment by jskeet** [comment]
Definitely, thanks. I strongly suspect this is going to be annoying in all kinds of ways, but we'll see what we can do...

**Comment by jskeet** [comment]
Just to add to the record, C# throws an exception too:

```csharp
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;

var fieldMask = new FieldMask
{
    Paths = { "user.display_name", "photo", "reviews.`John Smith`" }
};

Console.WriteLine(JsonFormatter.Default.Format(fieldMask));
```

Unhandled exception. System.InvalidOperationException: Invalid field mask to be converted to JSON: reviews.`John Smith`

Internal design doc started - but not much progress yet. (This will take a while, I'm afraid.)

**Comment by jskeet** [comment]
Just as a heads-up after a while, I'm coming back to FieldMask stuff at the moment. There's a lot to do, and I'm trying to put together a plan for what's required before executing on any of it. This will take quite a while, I'm afraid - but hopefully we'll get to a better place in the end.

**Comment by antongrbin** [comment]
Thank you for the heads up and thank you for working on this!

For transparency, I can share how my team mitigated the problem in the meantime. We went with the minimal change over https://google.aip.dev/161#map-fields where we add a constraint that map keys must **always** be quoted with backticks. We implemented FieldMask ser/de (for now only in Kotlin) which doesn't transform between snake and camel for quoted path elements.

Having said this, I would be interested to learn early if your plan ends up diverging from https://google.aip.dev/161#map-fields significantly. For example there might be an alternative where we use `[]` syntax to access map keys.
